<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Processing data · ABC-MK</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">ABC-MK</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li><a class="tocitem" href="../analytical/">Analytical estimations</a></li><li class="is-active"><a class="tocitem" href>Processing data</a><ul class="internal"><li><a class="tocitem" href="#Parsing-TGP-and-DGN-data-1"><span>Parsing TGP and DGN data</span></a></li><li><a class="tocitem" href="#Processing-muti-FASTA-files-1"><span>Processing muti-FASTA files</span></a></li></ul></li><li><a class="tocitem" href="../mk/">MK approaches</a></li></ul></li><li><span class="tocitem">Infering the rate and strength of adaptation</span><ul><li><a class="tocitem" href="../empirical/">Empirical estimation</a></li><li><a class="tocitem" href="../rates/">Rates</a></li><li><a class="tocitem" href="../input/">Input data</a></li><li><a class="tocitem" href="../summstat/">Summary statistics</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li></ul></li><li><a class="tocitem" href="../cli/">Command-Line interface</a></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Package Overview</a></li><li class="is-active"><a href>Processing data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Processing data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/Analytical.jl/blob/master/docs/src/data.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Parsing-genomic-data-1"><a class="docs-heading-anchor" href="#Parsing-genomic-data-1">Parsing genomic data</a><a class="docs-heading-anchor-permalink" href="#Parsing-genomic-data-1" title="Permalink"></a></h1><p>The module includes functions to parse TGP from Uricchio et al. (2019) and DGN from Murga-Moreno et al. (2019). In addition, the module have a function to parse SFS and divergence from multi-FASTA data following Murga-Moreno et al. (2019)</p><p>Please to parse raw data into SFS and divergence counts, first download raw files deposited in our repository:  </p><ul><li><a href="https://raw.githubusercontent.com/jmurga/Analytical.jl/master/data/tgp.txt">TGP</a>  </li><li><a href="https://raw.githubusercontent.com/jmurga/Analytical.jl/master/data/dgnRal.txt">DGN Zambia population</a>  </li><li><a href="https://raw.githubusercontent.com/jmurga/Analytical.jl/master/data/dgnZi.txt">DGN Raleigh population</a>  </li></ul><h2 id="Parsing-TGP-and-DGN-data-1"><a class="docs-heading-anchor" href="#Parsing-TGP-and-DGN-data-1">Parsing TGP and DGN data</a><a class="docs-heading-anchor-permalink" href="#Parsing-TGP-and-DGN-data-1" title="Permalink"></a></h2><p>Once you have downloaded the files, you can use the function <code>Analytical.parseSfs</code> to convert the data into SFS and divergence counts. Please check <a href="../reference/#Analytical.parseSfs"><code>Analytical.parseSfs</code></a> to get more info.</p><pre><code class="language-julia">alpha, sfs, divergence = Analytical.parseSfs(sampleSize = 661, data = &quot;/home/jmurga/tgpData/tgp.txt&quot;)</code></pre><p>To save the data, you can use CSV and DataFrames packages</p><pre><code class="language-julia">using CSV, DataFrames
CSV.write(&quot;/home/jmurga/tgpData/tgpSfs.tsv&quot;,DataFrame(sfs),delim=&#39;\t&#39;,header=false))
CSV.write(&quot;/home/jmurga/tgpData/tgpDiv.tsv&quot;,DataFrame(permutedims(divergence)),delim=&#39;\t&#39;,header=false))</code></pre><p>It is possible to directly subset genes IDs using Ensembl or Flybase id. Use a variable of type <code>Matrix{String,1}</code> into the argument <em>geneList</em></p><pre><code class="language-julia">ensemblList = CSV.read(&quot;/home/jmurga/tgpData/ensemblList.txt&quot;,header=false,DataFrame) |&gt; Array

alpha, sfs, divergence = Analytical.parseSfs(sampleSize = 661, data = &quot;/home/jmurga/tgpData/tgp.txt&quot;,geneList = ensemblList)</code></pre><p>If you are going to parse DGN, you need to change the value of the argument <em>isoline</em> to <em>true</em>. Following the Murga-Moreno et al. (2019) sample size for each population is:</p><ul><li>Zambia population: 154</li><li>RAL population: 160</li></ul><pre><code class="language-julia">alpha, sfs, divergence = Analytical.parseSfs(sampleSize = 160, data = &quot;/home/jmurga/dgnRal.txt&quot;,isolines=true)</code></pre><h2 id="Processing-muti-FASTA-files-1"><a class="docs-heading-anchor" href="#Processing-muti-FASTA-files-1">Processing muti-FASTA files</a><a class="docs-heading-anchor-permalink" href="#Processing-muti-FASTA-files-1" title="Permalink"></a></h2><p>We included some tools to process multi-FASTA files into unfolded SFS and divergence. The function <a href="@ref"><code>Analytical.uSfsFromFasta</code></a> needs three files: a reference file to degenerate the sequence and a multi-FASTA file to process the polymorphism, and an outgroup sequence to process the divergence. Consider downloading DGN data from <a href="https://www.johnpool.net/">John Pool lab</a> or <a href="https://popfly.uab.cat">PopFly</a> to obtain these files.</p><pre><code class="language-julia">&lt;!-- sfs, div = Analytical.uSfsFromFasta(
                        file = &quot;/home/jmurga/Downloads/example.fa&quot;,
                        reference  = &quot;/home/jmurga/Downloads/ref.fa&quot;,
                        outgroup   = &quot;/home/jmurga/Downloads/outgroups.fa&quot;,
                        samples    = 160,
                        bins       = 20,
                        codonTable = &quot;standard&quot;
                     ) --&gt;</code></pre><p>The scripts transforms a multi-FASTA file into a matrix which eliminates the monomorphic sites and process the divergent and polymorphic sites. Only 0-fold and 4-fold sites are analyzed as proxy of neutral and selected alleles. The following matrix correspond to the matrix processed at the above example.</p><pre><code class="language-julia">&lt;!-- 162×153456 Array{Char,2}:
 &#39;0&#39;  &#39;0&#39;  &#39;0&#39;  &#39;0&#39;  &#39;0&#39;  &#39;0&#39;  &#39;4&#39;  …  &#39;0&#39;  &#39;0&#39;  &#39;4&#39;  &#39;0&#39;  &#39;0&#39;  &#39;0&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;  …  &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 ⋮                        ⋮         ⋱  ⋮                        ⋮
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;  …  &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;A&#39;  &#39;C&#39;  &#39;C&#39;  &#39;T&#39;  …  &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39;
 &#39;A&#39;  &#39;A&#39;  &#39;G&#39;  &#39;C&#39;  &#39;C&#39;  &#39;A&#39;  &#39;T&#39;     &#39;G&#39;  &#39;G&#39;  &#39;C&#39;  &#39;A&#39;  &#39;G&#39;  &#39;T&#39; --&gt;</code></pre><p>Each column correspond to a sample position. Only fixed divergence and biallelic sites are process. The matrix is iterated through columns to discard the rest of the cases. The matrix is iterated to check the columns independently. In this way we check fixed divergence or biallelic polymorphism. For example at the following array all polymorphic positions (from row 2 to row 161) correspond to &#39;A&#39; nucleotides where as the outgroup (last row) correspond to &#39;G&#39;. This columns is addedas to neutral or selected divergence depedending on the first row value:</p><pre><code class="language-julia">&lt;!-- 162×1 Array{Char,1}:
 &#39;0&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 ⋮  
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;A&#39;
 &#39;C&#39; --&gt;</code></pre><p>The next example represent one polymorphic site taking into account the derived allele. At this array, the first 3 polymorphic positions correspond to &#39;G&#39; where as the rest of polymorphic sites correspond to C. Since &#39;C&#39; is the ancestral nucleotide (last row) the derived allele frequency will be 3/160:</p><pre><code class="language-julia">&lt;!-- 162×1 Array{Char,1}:
 &#39;4&#39;
 &#39;G&#39;
 &#39;G&#39;
 &#39;G&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 ⋮  
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39;
 &#39;C&#39; --&gt;</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../analytical/">« Analytical estimations</a><a class="docs-footer-nextpage" href="../mk/">MK approaches »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 11 May 2021 17:26">Tuesday 11 May 2021</span>. Using Julia version 1.6.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
